### ISSUE TITLE
Problem with data masking

### ISSUE THREAD

**Issue by melifluos** [arahmatiiii/E2N#1]
There seems to be an issue with your masks where you include every node in your train mask here

https://github.com/arahmatiiii/E2N/blob/main/preparation_graph_collab.py#L77

It looks like it should be:

TRAIN_MASK = torch.tensor([H.nodes[node]['mask'] == 'train' for node in H.nodes() **if node.mask == 'train'**], dtype=torch.bool)

this probably explains the performance improvements of your method, but if I'm wrong, then your method is extremely interesting.

@weihua916

**Comment by GlavitsBalazs** [comment]
@melifluos I believe that your suggestion is incorrect. 
`TRAIN_MASK = torch.tensor([H.nodes[node]['mask'] == 'train' for node in H.nodes() if node.mask == 'train'], dtype=torch.bool)` raises an AttributeError because in networkx `H.nodes()` is an iterable of ints so `node` is an int and it has no `mask` attribute.

`TRAIN_MASK = torch.tensor([H.nodes[node]['mask'] == 'train' for node in H.nodes() if H.nodes[node]['mask'] == 'train'], dtype=torch.bool)` would produce a tensor of shape `(1572069,)` with all elements set to True, which is not suitable for masking the node features.

However, I would like to raise a new question.

This line [here](https://github.com/arahmatiiii/E2N/blob/6644583551d24b267130161a74cd8315328edc60/train_collab.py#L23) concerns me.

```python
 graph_logits = MODEL(DATA.x, DATA.edge_index)
```
The entire dataset: training, validation, and testing splits included gets passed in to the model. Then only afterwards [in this line](https://github.com/arahmatiiii/E2N/blob/6644583551d24b267130161a74cd8315328edc60/train_collab.py#L30) are the testing samples separated for further evaluation.

```python
logit_ = graph_predictions[DATA.test_mask].cpu()
```

I suspect that this is a major data leakage issue. If true, this would warrant having the paper retracted.

I also find this line [here](https://github.com/arahmatiiii/E2N/blob/6644583551d24b267130161a74cd8315328edc60/train_collab.py#L74) suspicious.

```python
NEW_TRAIN_MASK = torch.zeros(DATA.edge_index.shape[1], dtype=torch.bool).to(args.device)
NEW_TRAIN_MASK[DATA.train_mask.nonzero()] = True
TRAIN_EDGE_INDEX = DATA.edge_index[:, NEW_TRAIN_MASK]
```
The code selects an edge at index `i` in the edge list as a training edge, on the condition that the node at the same index `i` in the node list is masked as a training node. I think that this can't be right, because there is no correspondence between the ordering of nodes in the node list and edges in the edge list. It just selects a few edges at random. The number of selected edges is also incorrect. It only selects one edge for every training node, but there should at least be two incident dummy edges for every dummy node.

I've found other strange things in the code as well. Such as [here](https://github.com/arahmatiiii/E2N/blob/6644583551d24b267130161a74cd8315328edc60/train_collab.py#L95), the optimizer is not set up to train the the node embeddings defined [above](https://github.com/arahmatiiii/E2N/blob/6644583551d24b267130161a74cd8315328edc60/train_collab.py#L88). And in model.py, the GCN only has one layer with 512 input channels and 3 output channels. Therefore the model only has 512*3+3=1539 trainable parameters, which is incredibly low.

**Comment by melifluos** [comment]
@GlavitsBalazs you are correct and I was not! Thank you for jumping in.

**Comment by arahmatiiii** [comment]
@GlavitsBalazs 

- To calculate the #Params I've used the `sum(p.numel() for p in model.parameters()). ` as the [leader_overview](https://ogb.stanford.edu/docs/leader_overview/)

- I think the Optimizer is correct according to this  [PyG Example](https://colab.research.google.com/drive/14OvFnAXggxB8vM4e8vSURUp1TaKnovzX?usp=sharing#scrollTo=pcr9joFQ6Mri) because the node embedding will be placed in the Data object and the Data will pass to the Model and then the Model will be Optimize

- My purpose in TRAIN_EDGE_INDEX was to split the train edges and nodes, then train the Model.  As I've just checked these two PyG official Node Classification Examples([PyG Example](https://colab.research.google.com/drive/14OvFnAXggxB8vM4e8vSURUp1TaKnovzX?usp=sharing#scrollTo=pcr9joFQ6Mri) and [PyG Example 2](https://colab.research.google.com/github/wandb/examples/blob/master/colabs/pyg/8_Node_Classification_(with_W&B).ipynb) ) I can retrain the model like this:


```
def train():
      model.train()
      optimizer.zero_grad()  # Clear gradients.
      out = model(data.x, data.edge_index)  # Perform a single forward pass.
      loss = criterion(out[data.train_mask], data.y[data.train_mask])  # Compute the loss solely based on the training nodes.
      loss.backward()  # Derive gradients.
      optimizer.step()  # Update parameters based on gradients.
      return loss

def test():
      model.eval()
      out = model(data.x, data.edge_index)
      pred = out.argmax(dim=1)  # Use the class with highest probability.
      test_correct = pred[data.test_mask] == data.y[data.test_mask]  # Check against ground-truth labels.
      test_acc = int(test_correct.sum()) / int(data.test_mask.sum())  # Derive ratio of correct predictions.
      return test_acc
```
I mean pass the whole `Data.x` and `Data.edge_index` to model in the train phase if this is correct?!

**Comment by arahmatiiii** [comment]
Thank you, @GlavitsBalazs, for your comments pointing out the concerns (possibility of data leakage). To satisfy this concern:

I've changed some parts making sure that there is no such possibility (please check the `preparation_graph_ddi.py` and `train_ddi.py`):

- In the `preparation_graph_ddi.py` I've added masks to each edge in H, denoting that each edge belongs to which one of the sets (train, test, or valid). According to these edge's masks, I have added the `EDGE_ATTR` to the `pyg_data` object.

- In the `train_ddi.py` (code below), we selects only edges that belong to the train set for training the model:

```
INDICES_WITH_ATTR_TRAIN = torch.where(DATA.edge_attr == EDGE_ATT_TO_NUMERIC['train'])[0]
TRAIN_EDGE_INDEX = DATA.edge_index[:, INDICES_WITH_ATTR_TRAIN]
```

We have about ~1.4m train nodes (train pos edges in G + train neg edges in G), as we add one dummy node and two dummy edges corresponding to each edge in G. Finally, we have about ~4.2m train edges (3*1.4m). This satisfies the concern, the model will be trained with only train edges.

- Additionally In the `train_ddi.py` I've added the `Data.x` to the Optimizer.
- I trained the ddi dataset with this modification and achieved the 99.5Hit@20 for the test set.

Your comments helped us to get even better performance for the ddi dataset; I am retraining for the collab and ppa datasets too.
I would appreciate it if you let me know any further ... (if exists)

**Comment by GlavitsBalazs** [comment]
Hi @arahmatiiii I'm testing [this revision](https://github.com/arahmatiiii/E2N/tree/b52d584516dbc3b2bd454bf2fe87a8444d53d754) of the repo because data/Raw/ddi is still available there.

Yep. The way you select `TRAIN_EDGE_INDEX` in `train_ddi.py` and `preparation_graph_ddi.py` looks correct now . Thanks for fixing it.

- I've noticed an obvious data leakage problem [in your code](https://github.com/arahmatiiii/E2N/blob/b52d584516dbc3b2bd454bf2fe87a8444d53d754/train_ddi.py#L103). 
    ```python
    train_logits = MODEL(DATA.x, TRAIN_EDGE_INDEX)
    train_loss = CRITERION(train_logits.cpu(), torch.tensor(DATA.label).cpu())
    ```

    You pass the testing node predictions and testing node labels to your training loss function.
    Those lines should be 
    ```python
    train_logits = MODEL(DATA.x, TRAIN_EDGE_INDEX)[DATA.train_mask]
    train_loss = CRITERION(train_logits.cpu(), torch.tensor(TRAIN_LABELS).cpu())
    ```

- I have an additional small suggestion to make. [This](https://github.com/arahmatiiii/E2N/blob/b52d584516dbc3b2bd454bf2fe87a8444d53d754/train_ddi.py#L117C22-L117C22):
    ```python
    BEST_OF_EACH_RUN_TEST.append(max(best_test))
    BEST_OF_EACH_RUN_VAL.append(best_val[best_test.index(max(best_test))])
    ```

    should be replaced with

    ```python
    BEST_OF_EACH_RUN_VAL.append(max(best_val))
    BEST_OF_EACH_RUN_TEST.append(best_test[best_val.index(max(best_val))])
    ```
    The testing set should not be used for early stopping. That's what the validation set is for. 
- I see that you've also added a new GCNConv layer in GCNBinaryNodeClassifier. You're up to 16899 parameters now, which is worth noting. Actually the number of trainable parameters is much larger now. There are 128 dimensional embeddings for every node: 242253696 + 16,899.

    This is a new data leakage concern. You've placed every node embedding into the optimizer for training. Testing nodes included. I suppose this is not really a problem, because `TRAIN_EDGE_INDEX` does not contain any testing nodes, therefore the gradients of the testing node embeddings should be 0.

    Just to be precise, `TRAIN_EDGE_INDEX` is incident with 1428148 nodes, so the embedding matrix has 182802944 truly trainable parameters.

I've ran the code with the modifications I propose here, and additionally I've removed `DATA.x` from the optimizer, just to be safe. It produced 99.9% test hits@20 after 50 epochs.

 I admit that I'm skeptical about these results. Honestly, it sounds too good to be true. I've tried to scrutinize your code to notice any mistakes, but at this point it eludes me.

**Comment by VishveshVsk** [comment]
@GlavitsBalazs What are your thoughts on this?
https://github.com/arahmatiiii/E2N/issues/3

**Comment by GlavitsBalazs** [comment]
@VishveshVsk I think you solved the mystery.

Everything mentioned so far in #1 is basically irrelevant.

**Comment by arahmatiiii** [comment]
Thank you, @VishveshVsk, for your comment.
I looked at the pdf of your implementation. If you see, you have used an old version of our preparation_graph_ddi code; 
the new version that updated then gives label 1 to neg and 2 for pos:
https://github.com/arahmatiiii/E2N/blob/f6c8a423833b31668966ae846e5842880d9ba928/preparation_graph_ddi.py#L58

Computing Hits@k looks fine to me.

**Comment by arahmatiiii** [comment]
Thank you, @GlavitsBalazs, for your nice comments making everything more clear.
First, as you see from my previous reply to @VishveshVsk, there is no problem with computing Hits@K!
Then, I've just applied your suggestions on `train_ddi.py` and also changed the `model.py` to include only two GCN layers. Similar to yours, I achieved 99.9Hits@20 on the ddi test set too. Like what you did, to be safe,`DATA.x` is removed from the optimizer, so we have 16,899 trainable parameters. I'll appreciate it if I hear more thoughts from you, if you have any ...

**Comment by VishveshVsk** [comment]
@arahmatiiii 
Which label is assigned to positive and negative edges does not seem to affect the network. Even when I assign label 2 to positive and 1 to negative, in y_pred_neg of the evaluator, the probability with which model predicts negative edges are positive has to be input and not the probability of negative edges being negative (which is low and should not be the case for a good performing model).
As I asked for, if my model training is inadequate, I encourage you to share the confusion matrix for the predictions of the model and we can resolve this.

**Comment by GlavitsBalazs** [comment]
I see that the suggestions I've made in this thread have made it into the code. Also the conversation has strayed far away from matters of data masking. @melifluos would you please close this issue now?

**Comment by melifluos** [comment]
As requested by @GlavitsBalazs
